src/server.rs

```diff
@@ -0,0 +1,213 @@
+        });
+
+        // Wait for listener
+        thread::sleep(Duration::from_millis(500));
+
+        let do_request = async move {
+            let request = Request::post(uri)
+                .header("Content-Type", "application/json")
+                .body(Body::from(req))
+                .unwrap();
+
+            Client::new().request(request).await
+        };
+
+        let response = threaded_rt.block_on(do_request)?;
+        let response_str: String = threaded_rt.block_on(body_to_string(response));
+
+        shutdown_sender.send(()).ok();
+
+        // Wait for shutdown
+        thread::sleep(Duration::from_millis(500));
+        threaded_rt.shutdown_background();
+
+        Ok(response_str)
+    }
+
+    /// Single hop to demonstrate request validation and onion unwrapping.
+    /// UTXO creation and bulletproof generation reserved for milestones 2 & 3.
+    #[test]
+    fn swap_lifecycle() -> Result<(), Box<dyn std::error::Error>> {
+        let server_key = secp::insecure_rand_secret()?;
+        
+        let secp = secp::Secp256k1::new();
+        let value: u64 = 100;
+        let blind = secp::insecure_rand_secret()?;
+        let commitment = secp::commit(value, &blind)?;
+        let session_key = secp::insecure_rand_secret()?;
+
+        let hop = types::Hop {
+            pubkey: secp::PublicKey::from_secret_key(&secp, &server_key)?,
+            payload: types::Payload{
+                excess: secp::insecure_rand_secret()?,
+                rangeproof: None,
+            }
+        };
+        let hops: Vec<types::Hop> = vec![hop];
+        let onion_packet = onion::create_onion(&commitment, &session_key, &hops)?;
+        let msg : Vec<u8> = vec![0u8, 1u8, 2u8, 3u8];
+        let comsig = secp::ComSignature::sign(value, &blind, &msg)?;
+        let swap = server::SwapReq{
+            onion: onion_packet,
+            msg: msg,
+            comsig: comsig,
+        };
+
+        let req = format!("{{\"jsonrpc\": \"2.0\", \"method\": \"swap\", \"params\": [{}], \"id\": \"1\"}}", serde_json::json!(swap));
+        let response = make_request(server_key, req)?;
+        let expected = "{\"jsonrpc\":\"2.0\",\"result\":\"success\",\"id\":\"1\"}\n";
+        assert_eq!(response, expected);
+        Ok(())
+    }
+
+    #[test]
+    fn swap_bad_request() -> Result<(), Box<dyn std::error::Error>> {
+        let params = "{ \"param\": \"Not a valid Swap request\" }";
+        let req = format!("{{\"jsonrpc\": \"2.0\", \"method\": \"swap\", \"params\": [{}], \"id\": \"1\"}}", params);
+        let response = make_request(secp::insecure_rand_secret()?, req)?;
+        let expected = "{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32602,\"message\":\"Invalid params: missing field `onion`.\"},\"id\":\"1\"}\n";
+        assert_eq!(response, expected);
+        Ok(())
+    }
+
+    // milestone 2 - add tests to cover invalid comsig's & inputs not in utxo set

```

Comment:
Another one we'll eventually want to add is to check that we accept only 1 swap for a given input - this way we close the attack vector when someone creates two swaps for the same input commitment. Similarly, the last node will want to check that no two swaps create the same output commitment.

Reply:
Multiple occurrences of the same input or output indicates either malice, a malfunctioning wallet, or multiple wallets sharing the same seed (which can be considered a form of malfunctioning too). In this case we can either drop all but one of the spends, or drop all of them.

Reply:
I vote for dropping all of them. This way we can see the malfunctioning frequency faster because honest users will complain.

Reply:
Wallets sharing the same seed is probably too common for us to just drop all duplicate inputs. We should decide between the earliest submission or the most recent one.

Duplicate outputs does seem more like a form of maliciousness or wallet malfunction. For that, it may be better to drop both. node n wouldn't know which one was submitted first anyway.

Reply:
If you keep one of two inputs sent from two seed-sharing wallets, then the wallet whose input you dropped will malfunction anyway when its input moves in a way it doesn't understand. So I'm inclined to agree with phyro, that it's better to drop all.

Reply:
When wallets restore from a seed they have no way of knowing what the last used derivation index was (they can guess from the last output belonging to it, but it's obviously not 100%), so non-malicious duplicate outputs can occur with non-zero frequency. The wallet stores an output's MMR position (once the output is found on-chain) to disambiguate it internally in this instance. It might be possible to make MMR position an optional field here, so duplicate outputs can be dropped if they're the same and neither has an MMR position, of if they both have the same MMR index. If only one provides an MMR pos it is preferred and other dropped. (note, I need to check whether wallet stores position or index)

Reply:
what scenario do you envision in which a user is hampered by duplicate invalidation?
does it involve a user mixing within one day both before and after restore?
btw, i'd rather not complicate the protocol with optional fields.

Reply:
Suppose you have 2 wallets sharing the same seed `w1` and `w2`. If the output was created with `w2`, then it's up to `w2` to create the coinswap. A wallet can know whether it created an output or not unless there has been a restore, but if a restore was done, you can just create a coinswap after 25 hours or so.